src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs

```diff
@@ -103,26 +99,91 @@ public override Expression VisitSqlFunction(SqlFunctionExpression sqlFunctionExp
+
+            case ExpressionType.ArrayIndex:
+                GenerateArrayIndex(expression);
+                return expression;
+            }
+
+            return base.VisitBinary(expression);
+        }
+
+        protected override Expression VisitUnary(UnaryExpression expression)
+        {
+            if (expression.NodeType == ExpressionType.ArrayLength)
             {
-                Sql.Append("(");
-                var exp = base.VisitBinary(binaryExpression);
-                Sql.Append(")");
-                return exp;
+                VisitSqlFunction(new SqlFunctionExpression("array_length", typeof(int), new[] { expression.Operand, Expression.Constant(1) }));
+                return expression;
             }
 
-            return base.VisitBinary(binaryExpression);
+            return base.VisitUnary(expression);
         }
 
+        public void GenerateArrayIndex([NotNull] BinaryExpression expression)
+        {
+            Debug.Assert(expression.NodeType == ExpressionType.ArrayIndex);
+
+            if (expression.Left.Type == typeof(byte[]))
+            {
+                // bytea cannot be subscripted, but there's get_byte
+                VisitSqlFunction(new SqlFunctionExpression("get_byte", typeof(byte),
+                    new[] { expression.Left, expression.Right }));
+                return;
+            }
+
+            if (expression.Left.Type == typeof(string))
+            {
+                // text cannot be subscripted, use substr
+                // PostgreSQL substr() is 1-based.
+
+                VisitSqlFunction(new SqlFunctionExpression("substr", typeof(char),
+                    new[] { expression.Left, expression.Right, Expression.Constant(1) }));
+                return;
+            }
+
+            // Regular array from here
+            Visit(expression.Left);
+            Sql.Append('[');
+            Visit(GenerateOneBasedIndexExpression(expression.Right));
+            Sql.Append(']');
+        }
+
+        public void GenerateArrayAny(ArrayAnyExpression arrayAnyExpression)
+        {
+            Visit(arrayAnyExpression.Operand);
+            Sql.Append(" = ANY (");
+            Visit(arrayAnyExpression.Array);
+            Sql.Append(")");
+        }
+
+        // PostgreSQL array indexing is 1-based. If the index happens to be a constant,
+        // just increment it. Otherwise, append a +1 in the SQL.
+        Expression GenerateOneBasedIndexExpression(Expression expression)
+            => expression is ConstantExpression constantExpression
+                ? Expression.Constant(Convert.ToInt32(constantExpression.Value) + 1)
+                : (Expression)Expression.Add(expression, Expression.Constant(1));
+
         // See http://www.postgresql.org/docs/current/static/functions-matching.html
-        public Expression VisitRegexMatch([NotNull] RegexMatchExpression regexMatchExpression)
+        public void GenerateRegexMatch([NotNull] RegexMatchExpression regexMatchExpression)
```

Comment:
Sorry, if my comment was not accurately constructed. Basically the pattern we follow is like this.
Whatever is invoked as a part of ReLinq Expression Visitor is made `Visit*` This also includes methods from `ISqlExpressionVisitor`. Hence its `VisitLike` instead of `GenerateLike`.

Easy way to spot is signature in form of `VisitXyz(XyzExpression expression)` Visit method visits particular type of expression. (Therefore VisitBinary method which visits BinaryExpression)

All `Generate*` methods are auxiliary method which are called explicit from inside of this class to handle particular case. Like `GenerateNotIn` which takes argument of `InExpression` (There is also `VisitIn` method which is invoked through ReLinq pipeline).

Since `GenerateArrayIndex` took parameter of type BinaryExpression, it should be named Generate but for this one, it should be `VisitRegexMatch`.

Hope this makes sense. Sorry for confusion.